File: backend/src/handlers/budgets.rs
// Copyright (c) 2024 Tempo Budget
// SPDX-License-Identifier: MIT
//
// Budget management HTTP handlers.

//! # Budget Handlers
//!
//! This module provides HTTP handlers for budget management:
//! - `GET /api/budgets` - List all budgets for the authenticated user
//! - `POST /api/budgets` - Create a new budget
//! - `GET /api/budgets/:id` - Get a specific budget
//! - `DELETE /api/budgets/:id` - Delete a budget

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use std::sync::Arc;
use uuid::Uuid;
use chrono::Utc;
use utoipa;

use crate::{
    auth::AuthUser,
    models::{Budget, CreateBudget},
    DbPool,
};
use crate::models::InviteMemberRequest;

/// Retrieves all budgets for the authenticated user.
///
/// # Endpoint
/// `GET /api/budgets`
///
/// # Authentication
/// Requires a valid JWT token in the Authorization header.
///
/// # Returns
/// - `200 OK` with array of `Budget` objects
/// - `500 Internal Server Error` if the query fails
#[utoipa::path(
    get,
    path = "/api/budgets",
    tag = "budgets",
    responses(
        (status = 200, description = "List of user's budgets", body = Vec<Budget>),
        (status = 500, description = "Failed to fetch budgets")
    ),
    security(("bearer_auth" = []))
)]
pub async fn get_budgets(
    State(pool): State<Arc<DbPool>>,
    auth: AuthUser,
) -> Result<Json<Vec<Budget>>, StatusCode> {
    let budgets = sqlx::query_as::<_, Budget>(
        "SELECT id, user_id, name, budget_type, is_active, created_at, updated_at
         FROM budgets WHERE user_id = ?"
    )
        .bind(&auth.user_id)
        .fetch_all(pool.as_ref())
        .await
        .map_err(|e| {
            tracing::error!("Failed to fetch budgets: {:?}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok(Json(budgets))
}

/// Creates a new budget.
///
/// # Endpoint
/// `POST /api/budgets`
///
/// # Authentication
/// Requires a valid JWT token in the Authorization header.
///
/// # Request Body
/// - `name`: Budget name/title
/// - `budget_type`: Either "personal" or "group"
///
/// # Returns
/// - `200 OK` with the created `Budget` object
/// - `500 Internal Server Error` if creation fails
///
/// # Notes
/// For group budgets, the creator is automatically added as the owner.
#[utoipa::path(
    post,
    path = "/api/budgets",
    tag = "budgets",
    request_body = CreateBudget,
    responses(
        (status = 200, description = "Budget created successfully", body = Budget),
        (status = 500, description = "Failed to create budget")
    ),
    security(("bearer_auth" = []))
)]
pub async fn create_budget(
    State(pool): State<Arc<DbPool>>,
    auth: AuthUser,
    Json(payload): Json<CreateBudget>,
) -> Result<Json<Budget>, StatusCode> {
    let id = Uuid::new_v4().to_string();
    let user_id = auth.user_id.clone();
    let now = Utc::now().to_rfc3339();

    sqlx::query(
        "INSERT INTO budgets (id, user_id, name, budget_type, is_active, created_at, updated_at)
         VALUES (?, ?, ?, ?, 0, ?, ?)"
    )
        .bind(&id)
        .bind(&user_id)
        .bind(&payload.name)
        .bind(&payload.budget_type)
        .bind(&now)
        .bind(&now)
        .execute(pool.as_ref())
        .await
        .map_err(|e| {
            tracing::error!("Failed to insert budget: {:?}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    // For group budgets, add the creator as owner
    if payload.budget_type == "group" {
        let member_id = Uuid::new_v4().to_string();
        sqlx::query(
            "INSERT INTO budget_members (id, budget_id, user_id, role, created_at) VALUES (?, ?, ?, 'owner', ?)"
        )
            .bind(&member_id)
            .bind(&id)
            .bind(&user_id)
            .bind(&now)
            .execute(pool.as_ref())
            .await
            .map_err(|e| {
                tracing::error!("Failed to insert budget member: {:?}", e);
                StatusCode::INTERNAL_SERVER_ERROR
            })?;
    }

    let budget = sqlx::query_as::<_, Budget>(
        "SELECT id, user_id, name, budget_type, is_active, created_at, updated_at
         FROM budgets WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|e| {
            tracing::error!("Failed to fetch created budget: {:?}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok(Json(budget))
}

/// Retrieves a specific budget by ID.
///
/// # Endpoint
/// `GET /api/budgets/:id`
///
/// # Path Parameters
/// - `id`: The budget's unique identifier
///
/// # Returns
/// - `200 OK` with the `Budget` object
/// - `404 Not Found` if the budget doesn't exist
#[utoipa::path(
    get,
    path = "/api/budgets/{id}",
    tag = "budgets",
    params(
        ("id" = String, Path, description = "Budget unique identifier")
    ),
    responses(
        (status = 200, description = "Budget details", body = Budget),
        (status = 404, description = "Budget not found")
    )
)]
pub async fn get_budget(
    State(pool): State<Arc<DbPool>>,
    Path(id): Path<String>,
) -> Result<Json<Budget>, StatusCode> {
    let budget = sqlx::query_as::<_, Budget>(
        "SELECT id, user_id, name, budget_type, is_active, created_at, updated_at
         FROM budgets WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|e| {
            tracing::error!("Failed to fetch budget: {:?}", e);
            StatusCode::NOT_FOUND
        })?;

    Ok(Json(budget))
}

/// Deletes a budget.
///
/// # Endpoint
/// `DELETE /api/budgets/:id`
///
/// # Authentication
/// Requires a valid JWT token in the Authorization header.
///
/// # Path Parameters
/// - `id`: The budget's unique identifier
///
/// # Returns
/// - `204 No Content` on successful deletion
/// - `403 Forbidden` if the user doesn't own the budget
/// - `500 Internal Server Error` if deletion fails
///
/// # Authorization
/// User must be either the budget owner (for group budgets) or the creator
/// (for personal budgets).
#[utoipa::path(
    delete,
    path = "/api/budgets/{id}",
    tag = "budgets",
    params(
        ("id" = String, Path, description = "Budget unique identifier")
    ),
    responses(
        (status = 204, description = "Budget deleted successfully"),
        (status = 403, description = "Not authorized to delete this budget"),
        (status = 500, description = "Failed to delete budget")
    ),
    security(("bearer_auth" = []))
)]
pub async fn delete_budget(
    State(pool): State<Arc<DbPool>>,
    auth: AuthUser,
    Path(id): Path<String>,
) -> Result<StatusCode, StatusCode> {
    // Check if user is owner (for group budgets)
    let is_owner = sqlx::query_scalar::<_, i64>(
        "SELECT COUNT(*) FROM budget_members WHERE budget_id = ? AND user_id = ? AND role = 'owner'"
    )
        .bind(&id)
        .bind(&auth.user_id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    // Check if it's the user's own budget (for personal budgets)
    let is_my_budget = sqlx::query_scalar::<_, i64>(
        "SELECT COUNT(*) FROM budgets WHERE id = ? AND user_id = ?"
    )
        .bind(&id)
        .bind(&auth.user_id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    if is_owner == 0 && is_my_budget == 0 {
        return Err(StatusCode::FORBIDDEN);
    }

    // Delete the budget (cascade deletes related records)
    sqlx::query("DELETE FROM budgets WHERE id = ?")
        .bind(&id)
        .execute(pool.as_ref())
        .await
        .map_err(|e| {
            tracing::error!("Failed to delete budget: {:?}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok(StatusCode::NO_CONTENT)
}

File: backend/app/routes/budgets.py
# Copyright (c) 2024 Tempo Budget
# SPDX-License-Identifier: MIT
"""Budget management routes."""

from datetime import datetime, timezone
from typing import Annotated
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from ..auth import get_current_user
from ..database import get_db
from ..models import Budget, CreateBudget, BudgetSummary

router = APIRouter()


@router.get("", response_model=list[Budget])
async def get_budgets(
    user_id: Annotated[str, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db)
):
    """Retrieve all budgets for the authenticated user (owned + shared)."""
    result = await db.execute(
        text("""
            SELECT DISTINCT b.id, b.user_id, b.name, b.budget_type, b.is_active, b.created_at, b.updated_at
            FROM budgets b
            LEFT JOIN budget_members bm ON b.id = bm.budget_id
            WHERE b.user_id = :user_id OR bm.user_id = :user_id
        """),
        {"user_id": user_id}
    )
    rows = result.fetchall()
    return [Budget(
        id=row.id,
        user_id=row.user_id,
        name=row.name,
        budget_type=row.budget_type,
        is_active=row.is_active,
        created_at=row.created_at,
        updated_at=row.updated_at,
    ) for row in rows]


@router.post("", response_model=Budget)
async def create_budget(
    payload: CreateBudget,
    user_id: Annotated[str, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db)
):
    """Create a new budget."""
    budget_id = str(uuid4())
    now = datetime.now(timezone.utc).isoformat()

    await db.execute(
        text("""
            INSERT INTO budgets (id, user_id, name, budget_type, is_active, created_at, updated_at)
            VALUES (:id, :user_id, :name, :budget_type, 0, :created_at, :updated_at)
        """),
        {
            "id": budget_id,
            "user_id": user_id,
            "name": payload.name,
            "budget_type": payload.budget_type,
            "created_at": now,
            "updated_at": now,
        }
    )

    # For group budgets, add creator as owner
    if payload.budget_type == "group":
        member_id = str(uuid4())
        await db.execute(
            text("""
                INSERT INTO budget_members (id, budget_id, user_id, role, created_at)
                VALUES (:id, :budget_id, :user_id, 'owner', :created_at)
            """),
            {
                "id": member_id,
                "budget_id": budget_id,
                "user_id": user_id,
                "created_at": now,
            }
        )

    await db.commit()

    result = await db.execute(
        text("SELECT id, user_id, name, budget_type, is_active, created_at, updated_at FROM budgets WHERE id = :id"),
        {"id": budget_id}
    )
    row = result.fetchone()
    return Budget(
        id=row.id,
        user_id=row.user_id,
        name=row.name,
        budget_type=row.budget_type,
        is_active=row.is_active,
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


@router.get("/summaries", response_model=list[BudgetSummary])
async def get_budget_summaries(
    user_id: Annotated[str, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db)
):
    """Get summary statistics for all user budgets."""
    # Get all budgets for the user (owned + group member)
    result = await db.execute(
        text("""
            SELECT DISTINCT b.id, b.name, b.budget_type
            FROM budgets b
            LEFT JOIN budget_members bm ON b.id = bm.budget_id
            WHERE b.user_id = :user_id OR bm.user_id = :user_id
        """),
        {"user_id": user_id}
    )
    budgets = result.fetchall()

    summaries = []
    for budget in budgets:
        # Get total budget (sum of parent categories only)
        result = await db.execute(
            text("""
                SELECT COALESCE(SUM(amount), 0) as total
                FROM categories
                WHERE budget_id = :budget_id AND parent_id IS NULL
            """),
            {"budget_id": budget.id}
        )
        total_budget = result.fetchone().total

        # Get total spent (expenses)
        result = await db.execute(
            text("""
                SELECT COALESCE(SUM(amount), 0) as total
                FROM transactions
                WHERE budget_id = :budget_id AND transaction_type = 'expense'
            """),
            {"budget_id": budget.id}
        )
        total_spent = result.fetchone().total

        # Get total income
        result = await db.execute(
            text("""
                SELECT COALESCE(SUM(amount), 0) as total
                FROM transactions
                WHERE budget_id = :budget_id AND transaction_type = 'income'
            """),
            {"budget_id": budget.id}
        )
        total_income = result.fetchone().total

        # Get category count
        result = await db.execute(
            text("SELECT COUNT(*) as cnt FROM categories WHERE budget_id = :budget_id"),
            {"budget_id": budget.id}
        )
        category_count = result.fetchone().cnt

        # Get transaction count
        result = await db.execute(
            text("SELECT COUNT(*) as cnt FROM transactions WHERE budget_id = :budget_id"),
            {"budget_id": budget.id}
        )
        transaction_count = result.fetchone().cnt

        remaining = total_budget - total_spent
        percentage = (total_spent / total_budget * 100) if total_budget > 0 else 0

        summaries.append(BudgetSummary(
            id=budget.id,
            name=budget.name,
            budget_type=budget.budget_type,
            total_budget=round(total_budget, 2),
            total_spent=round(total_spent, 2),
            total_income=round(total_income, 2),
            remaining=round(remaining, 2),
            percentage=round(percentage, 2),
            category_count=category_count,
            transaction_count=transaction_count,
        ))

    return summaries


@router.get("/{budget_id}", response_model=Budget)
async def get_budget(budget_id: str, db: AsyncSession = Depends(get_db)):
    """Retrieve a specific budget by ID."""
    result = await db.execute(
        text("SELECT id, user_id, name, budget_type, is_active, created_at, updated_at FROM budgets WHERE id = :id"),
        {"id": budget_id}
    )
    row = result.fetchone()

    if not row:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Budget not found")

    return Budget(
        id=row.id,
        user_id=row.user_id,
        name=row.name,
        budget_type=row.budget_type,
        is_active=row.is_active,
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


@router.delete("/{budget_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_budget(
    budget_id: str,
    user_id: Annotated[str, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db)
):
    """Delete a budget."""
    # Check if user is owner (for group budgets)
    result = await db.execute(
        text("SELECT COUNT(*) as cnt FROM budget_members WHERE budget_id = :budget_id AND user_id = :user_id AND role = 'owner'"),
        {"budget_id": budget_id, "user_id": user_id}
    )
    is_owner = result.fetchone().cnt > 0

    # Check if it's the user's own budget (for personal budgets)
    result = await db.execute(
        text("SELECT COUNT(*) as cnt FROM budgets WHERE id = :id AND user_id = :user_id"),
        {"id": budget_id, "user_id": user_id}
    )
    is_my_budget = result.fetchone().cnt > 0

    if not is_owner and not is_my_budget:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized")

    await db.execute(text("DELETE FROM budgets WHERE id = :id"), {"id": budget_id})
    await db.commit()
